src/recipes.zig

```diff
@@ -0,0 +1,217 @@
+					return error.NoMatch;
+				}
+				idx += literal.len;
+			},
+			.symbol => |symbol| {
+				const endIndex: usize = if(i + 1 < pattern.len) blk: {
+					const nextSegment = pattern[i + 1];
+					if(nextSegment == .symbol) {
+						return error.NoMatch;
+					}
+					break :blk std.mem.indexOfPos(u8, target, idx, nextSegment.literal) orelse {
+						return error.NoMatch;
+					};
+				} else target.len;
+				newKeys.put(allocator.dupe(u8, symbol), allocator.dupe(u8, target[idx..endIndex])) catch unreachable;
+				idx = endIndex;
+			},
+		}
+	}
+	if(idx == target.len) {
+		return newKeys;
+	} else {
+		return error.NoMatch;
+	}
+}
+
+const ItemKeyPair = struct {item: ItemStack, keys: std.StringHashMap([]const u8)};
+
+fn parseRecipeItem(allocator: NeverFailingAllocator, zon: ZonElement, keys: *const std.StringHashMap([]const u8)) !main.List(ItemKeyPair) {
+	var arenaAllocator: NeverFailingArenaAllocator = .init(main.stackAllocator);
+	defer arenaAllocator.deinit();
+	const arena = arenaAllocator.allocator();
+	var id = zon.as([]const u8, "");
+	id = std.mem.trim(u8, id, &std.ascii.whitespace);
+	var amount: u16 = 1;
+	if(std.mem.indexOfScalar(u8, id, ' ')) |index| blk: {
+		amount = std.fmt.parseInt(u16, id[0..index], 0) catch break :blk;
+		id = id[index + 1 ..];
+		id = std.mem.trim(u8, id, &std.ascii.whitespace);
+	}
+	var itemPairs: main.List(ItemKeyPair) = .initCapacity(allocator, 1);
+	const pattern = try parsePattern(arena, id, keys);
+	if(id.len > 0 and pattern.items.len == 1 and pattern.items[0] == .literal) {
+		const item = BaseItemIndex.fromId(pattern.items[0].literal) orelse return itemPairs;
+		itemPairs.append(.{
+			.item = .{
+				.item = .{.baseItem = item},
+				.amount = amount,
+			},
+			.keys = keys.clone() catch unreachable,
+		});
+	} else {
+		var iter = items.iterator();
+		while(iter.next()) |item| {
+			if(matchWithKeys(allocator, item.id(), pattern.items, keys) catch null) |newKeys| {
+				itemPairs.append(.{
+					.item = .{
+						.item = .{.baseItem = item.*},
+						.amount = amount,
+					},
+					.keys = newKeys,
+				});
+			}
+		}
+	}
+	return itemPairs;
+}
+
+fn generateItemCombos(allocator: NeverFailingAllocator, recipe: []ZonElement) !main.List([]ItemStack) {
+	var arenaAllocator: NeverFailingArenaAllocator = .init(main.stackAllocator);
+	defer arenaAllocator.deinit();
+	var arena = arenaAllocator.allocator();
+	var emptyKeys: std.StringHashMap([]const u8) = .init(arena.allocator);
+	const startingParsedItems = try parseRecipeItem(arena, recipe[0], &emptyKeys);
+	var inputCombos: main.List([]ItemStack) = .initCapacity(arena, startingParsedItems.items.len);
+	var keyList: main.List(std.StringHashMap([]const u8)) = .initCapacity(arena, startingParsedItems.items.len);
+	for(startingParsedItems.items) |item| {
+		const inputs = arena.alloc(ItemStack, recipe.len);
+		inputs[0] = item.item;
+		inputCombos.append(inputs);
+		keyList.append(item.keys);
+	}
```

Comment:
Instead of special-casing the first iteration, I think it would be simpler to just start the iteration with one empty inputCombo and one empty map in the keyList.